### PR TITLE
MLT writer perf improvements

### PR DESCRIPTION
**Pull request by jbachorik**
Another set of changes which are results of running the spring-petclinic benchmark and observing the collected profiles.

- the biggest win is the thread-local pooling of ByteBuffer which brings down the observed allocation rate from >4GiB/s to ~400MiB/s
- another 'free' improvement is realization that I used 'array' based fastutil maps and sets and this made each and every lookup insanely expensive; changing to hash-based version brings ~10x speedup in the stack sampling routine
- another change is rather marginal - I was experimenting with completely disabling the sub-tree deduplication and it seems like the collected data size is almost unaffected while the perf is slightly (<5%) improved; the default behaviour is not to attempt sub-tree deduplication but can be changed by setting the system property `mlt.subtree_compression` to `true` (for benchmarking/testing purposes mostly)

### CODE REVIEW

**Comment by richardstartin** [comment]
Since it looks like we dictionary encode stack frames and dictionary codes will usually be assigned sequentially to stack frames, can we delta encode stack traces prior to var int encoding? I suspect that most stack traces we capture will contain common sequences of stack frames.

File: dd-java-agent/agent-mlt/src/main/java/com/datadog/mlt/sampler/ScopeStackCollector.java [hunk 1]
```diff
@@ -148,8 +179,8 @@ void addCompressedStackptr(int stackptr) {
     stacks.add(stackptr);
```
Comment (richardstartin) [inline dd-java-agent/agent-mlt/src/main/java/com/datadog/mlt/sampler/ScopeStackCollector.java:179]:
```java
if ((topItem & 0x80000000) == 0x80000000
```
-> 

```java
if (topItem < 0)
```

https://github.com/DataDog/dd-trace-java/pull/1599/files#diff-a5e10a778b56aa8ff92e25ed4fd67ae4L121

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128ByteBufferWriter.java [hunk 2]
```diff
@@ -127,11 +158,31 @@ private void ensureCapacity(int offset, int dataLength) {
...
+      buffer = BUFFER_REF.get().get();
+      if (buffer == null) {
+        /*
+         * The underlying ByteBuffer was released due to memory pressure.
+         * Clear the TLS variable and try recreating the buffer.
+         */
+        log.info(
+            "Cleaning dangling reference to ByteBuffer from thread {}",
+            Thread.currentThread().getName());
+        BUFFER_REF.remove();
```
Comment (richardstartin) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128ByteBufferWriter.java:183]: I think we might need to cast to `DirectBuffer` and get access to the `Cleaner` unfortunately, which doesn't work on JDK11+, if we're going to use direct memory I think we need multi release facilities to do things like cleaning up.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128ByteBufferWriter.java [hunk 3]
```diff
@@ -127,11 +158,31 @@ private void ensureCapacity(int offset, int dataLength) {
...
-      buffer = newBuffer;
+      BUFFER_REF.set(new SoftReference<>(newBuffer));
+      return newBuffer;
     }
+    return buffer;
   }
 
   private static ByteBuffer allocateBuffer(int capacity) {
-    return ByteBuffer.allocateDirect(capacity);
+    return ByteBuffer.allocate(capacity);
```
Comment (richardstartin) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128ByteBufferWriter.java:168]: good.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java [hunk 4]
```diff
@@ -44,23 +39,16 @@ private void writeChunk(IMLTChunk chunk, LEB128Writer writer) {
...
-     */
-    LEB128Writer stackEventWriter = frameStackDataWriter;
     int[] eventCount = new int[1];
     chunk
         .frameSequenceCpIndexes()
         .forEach(
             val -> {
               eventCount[0]++;
-              stackEventWriter.writeInt(val);
               if ((val & 0x80000000) == 0) {
```
Comment (richardstartin) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java:52]:
```java

if (val > 0)
```

File: dd-java-agent/agent-mlt/src/main/java/com/datadog/mlt/sampler/ScopeStackCollector.java [hunk 5]
```diff
@@ -55,20 +65,37 @@ public void collect(StackTraceElement[] stackTrace) {
...
+                    element.getMethodName(),
+                    element.getLineNumber(),
+                    stringPool,
+                    framePool),
+                tree);
+      }
+    } else {
+      int[] framePtrs = new int[stackTrace.length];
+      for (int i = 0; i < stackTrace.length; i++) {
+        StackTraceElement element = stackTrace[i];
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply dd-java-agent/agent-mlt/src/main/java/com/datadog/mlt/sampler/ScopeStackCollector.java:85]: Yep. The former one will be probably removed - does not seem to bring much benefit in real life.

File: dd-java-agent/agent-mlt/src/test/java/com/datadog/mlt/sampler/ScopeManagerTest.java [hunk 6]
```diff
@@ -51,32 +50,37 @@ void sample() throws Exception {
       sampler.collect(ti.getStackTrace());
       Thread.sleep(ThreadLocalRandom.current().nextInt(20) + 2);
     }
-    IMLTChunk collectedChunk = sampler.end();
+    sampler.end(
+        collectedChunk -> {
+          byte[] serializedChunk = collectedChunk.serialize();
+          List<MLTChunk> chunks = new MLTReader().readMLTChunks(serializedChunk);
```
Comment (jpbempel) [inline dd-java-agent/agent-mlt/src/test/java/com/datadog/mlt/sampler/ScopeManagerTest.java:56]: MLTReader class does not have any fields so we may make all methods static.

usually when you are doing something like:
``` new MyClass().aMethod() ```
you can transform your method as a static one as don't have to store any fields inside the object

File: dd-java-agent/agent-mlt/src/main/java/com/datadog/mlt/sampler/JMXSampler.java [hunk 7]
```diff
@@ -25,7 +25,8 @@
   public JMXSampler(ThreadScopeMapper threadScopeMapper) {
     this.threadScopeMapper = threadScopeMapper;
     // TODO period as parameter
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply dd-java-agent/agent-mlt/src/main/java/com/datadog/mlt/sampler/JMXSampler.java:27]: It is not a proper parameter, though :)

File: dd-java-agent/agent-mlt/src/test/java/com/datadog/mlt/sampler/ScopeManagerTest.java (hunk 6 above)
Comment (jbachorik) [reply dd-java-agent/agent-mlt/src/test/java/com/datadog/mlt/sampler/ScopeManagerTest.java:56]: Yep, once the overall shape is stabilized I might do this.

**Pull request merged**